Improve initialization functions

mariaschuld opened the pull request:
*Note: This PR forks from PR #409, which should be reviewed first.*

**Context:** Another PR in the templates refactor series. Resolves Issue #383, by changing the signature of initialization functions, adding more functions and improving the docs.

**Description of the Change:** 

* Every template now has initialization functions ``_uniform`` and ``_normal`` for every parameter array it takes. For example, ``Interferometer`` has three initializations for its three inputs ``theta``, ``phi`` ``varphi``.   
* Added ``_all`` initializations for ``Interferometer`` and ``CVNeuralNetLayers`` to initialize all parameters in one go.
* Added new functions to introduction in docs, and gave code examples.

**Benefits:** Hopefully resolves confusion around usage, improves explanations for the user, sets a clear standard of how new templates are treated.

**Possible Drawbacks:** Adds quite a lot of functions to the ``init`` module, with a lot of code repetition. The justification is that these functions are for convenience, and should be as simple as possible (which was why we previously decided for ``_normal`` and ``_uniform`` versions, rather than keyword arguments).

**Related GitHub Issues:** #383

josh146 commented on doc/introduction/templates.rst:
```diff
@@ -159,10 +231,23 @@ randomly initialized arrays for the trainable parameters.
+    ~pennylane.init.cvqnn_layers_phi_a_uniform
+    ~pennylane.init.cvqnn_layers_phi_a_normal
+    ~pennylane.init.cvqnn_layers_phi_a_uniform
+    ~pennylane.init.cvqnn_layers_phi_a_normal
```
😆

josh146 commented on pennylane/__init__.py:
```diff
@@ -28,6 +28,7 @@
 import pennylane.init
 import pennylane.templates.layers
 import pennylane.templates.embeddings
+import pennylane.templates.subroutines
```
Will all layers, embeddings, and subroutines be imported here, so that you can do `qml.templates.StronglyEntangling` etc?

josh146 commented on pennylane/init.py:
```diff
@@ -33,22 +33,22 @@ def strong_ent_layers_uniform(n_layers, n_wires, low=0, high=2 * pi, seed=None):
 
     Returns:
-        list of parameter array
+        parameter array
```
```suggestion
        array: parameter array
```

josh146 commented on pennylane/init.py:
```diff
@@ -144,7 +92,7 @@ def random_layers_uniform(n_layers, n_wires, n_rots=None, low=0, high=2 * pi, se
 
     Returns:
-        list of parameter array
+        parameter array
```
```suggestion
        array: parameter array
```

josh146 commented on pennylane/init.py:
```diff
@@ -183,273 +131,586 @@ def random_layers_normal(n_layers, n_wires, n_rots=None, mean=0, std=0.1, seed=N
+    phi_a = cvqnn_layers_phi_a_uniform(**kwargs)
+    k = cvqnn_layers_kappa_normal(**kwargs)
 
+    return [theta_1, phi_1, varphi_1, r, phi_r, theta_2, phi_2, varphi_2, a, phi_a, k]
```
Nice!

josh146 commented on pennylane/init.py:
```diff
@@ -183,273 +131,586 @@ def random_layers_normal(n_layers, n_wires, n_rots=None, mean=0, std=0.1, seed=N
+            seed (int): seed used in sampling the parameters, makes function call deterministic
+
+        Returns:
+            parameter array
```
just in general need to add return types

josh146 commented on tests/test_init.py:
```diff
@@ -17,542 +17,194 @@
+        in correct interval."""
+        p = init(n_wires=n_wires, mean=1, std=0, seed=seed)
+        p_mean = np.mean(p)
+        assert np.isclose(p_mean, 1, atol=tol)
```
Should we have integration tests in this file. That is, check that the generated parameters are accepted by the templates

josh146 commented on tests/test_templates.py:
```diff
@@ -12,106 +12,256 @@
 """
-Unit tests for the integration of the methods from :mod:`pennylane.init` with the :mod:`pennylane.templates` module.
+Integration tests for templates, including integration with initialization functions
+in :mod:`pennylane.init`, running templates in larger circuits,
```
oh, integration tests are here 😆

josh146 commented on tests/test_templates.py:
```diff
@@ -12,106 +12,256 @@
+* When adding a new interface, try to import it and extend the fixture ``interfaces``.
+
+* When adding a new template, extend the fixtures ``qubit_const`` or ``cv_const`` by a *list* of arguments to the
+template. Note: Even if the template takes only one argument, it has to be wrapped in a list (i.e. [weights]).
```
Will this 'double nesting' work okay with TF?

josh146 commented on tests/test_templates.py:
```diff
@@ -12,106 +12,256 @@
 
-        circuit(weights=p)
+    if tf.__version__[0] == "1":
+        print(tf.__version__)
```
```suggestion
```

josh146 commented on .github/CHANGELOG.md:
```diff
@@ -11,8 +12,16 @@
 
+* Initialization functions now return single array of weights per function. Utility for multi-weight templates
+  `Interferometer()` and `CVNeuralNetLayers()` are provided.
+  [#XXX](https://github.com/XanaduAI/pennylane/pull/XXX)
```
```suggestion
  [#412](https://github.com/XanaduAI/pennylane/pull/412)
```

mariaschuld commented:
This one is now ready to (re-)review. It contains - and partially overrides - changes from #409.

mariaschuld commented:
Again ready for re-review...

co9olguy commented on tests/test_init.py:
```diff
@@ -17,542 +17,148 @@
+# Shapes
+
+SHP = [(repeat, n_rots), (repeat, n_wires, 3), intf, intf, wrs, wrs,
+       wrs, wrs, wrs, wrs, (n_if,), (n_if,), (n_wires,),
```
this one was hard to parse 😄

co9olguy replied:
Any way we can make this part clearer to people who might read and/or modify these tests?

co9olguy replied:
One suggestion: just using full words (i.e., `wires` instead of `wrs`) might make the cognitive load slightly easier

co9olguy commented on tests/test_init.py:
```diff
@@ -17,542 +17,148 @@
+##################
+
+
+class TestInitRepeated:
```
I 😍 how much code this replaced.

co9olguy replied:
I really like these new tests, they are compact and their intention is clear. 👍 
However, I think the parametrize arguments could be make a bit more readable, specifically `SHP`, `ALL_SHP`, `SHAPES`, and `ALL_SHAPES`

co9olguy commented on tests/test_init.py:
```diff
@@ -17,542 +17,148 @@
+        """Confirm that initialization functions
+         return an array with the correct shape."""
+        s = sgntr.copy()
+        s['seed'] = seed
```
perhaps instead of copying and updating the dict, we could just create a new dict which adds the seed?
Something like:
```
kwargs = {**s, 'seed':seed}
p = init(**kwargs)
```

co9olguy replied:
(applies throughout)

co9olguy commented on tests/test_init.py:
```diff
@@ -17,542 +17,148 @@
+        s['seed'] = seed
+
+        # Case A: Uniformly distributed parameters
+        if 'low' in sgntr.keys() and 'high' in s.keys():
```
why check for `low` in `sgntr` but then `high` in `s`?

co9olguy commented on pennylane/init.py:
```diff
@@ -183,321 +131,643 @@ def random_layers_normal(n_layers, n_wires, n_rots=None, mean=0, std=0.1, seed=N
+def cvqnn_layers_varphi_normal(n_layers, n_wires, mean=0, std=0.1, seed=None):
+    r"""Creates a parameter array for the ``varphi`` input to the interferometers of :func:`~.CVNeuralNetLayers`.
+
+        The parameters are drawn from a uniform distribution.
```
```suggestion
        The parameters are drawn from a normal distribution.
```

co9olguy commented on pennylane/init.py:
```diff
@@ -183,321 +131,643 @@ def random_layers_normal(n_layers, n_wires, n_rots=None, mean=0, std=0.1, seed=N
 
     Returns:
-        list of parameter arrays
+        list: list of parameter arrays
```
How come changed here but not in `interferometer_all` or `interferometer_theta_uniform`?

co9olguy replied:
In fact, the docstrings of these functions each say they create `a parameter array`. @mariaschuld can you double check?

co9olguy commented on pennylane/init.py:
```diff
@@ -183,321 +131,643 @@ def random_layers_normal(n_layers, n_wires, n_rots=None, mean=0, std=0.1, seed=N
 
     Returns:
-        list of parameter arrays
+        list: list of parameter arrays
```
see above comment

co9olguy commented on pennylane/init.py:
```diff
@@ -183,321 +131,643 @@ def random_layers_normal(n_layers, n_wires, n_rots=None, mean=0, std=0.1, seed=N
+        Returns:
+            array: parameter array
+        """
     if seed is not None:
```
*almost* feels like we should make a simple helper function do avoid a lot of duplicate code. Something like
```
def _sample_uniform(low, high, size, seed=None):
    if seed is not None:
        np.random.seed(seed)
    return np.random.uniform(low=low, high=high, size=size)
```

co9olguy replied:
(more maintainable too, since absolutely every `init` function would be fed through the same 2 helper functions)

mariaschuld commented:
Thanks @co9olguy ! I addressed all your comments, except from the one avoiding code repetition which I added to the issue on further template test improvements. 

Ready for re-review.

Note: Some of your comments were related to changes PR#409. I don't mind, but just don't get confused if they are updated on this branch but not in the original PR.

mariaschuld commented on .github/CHANGELOG.md:
```diff
@@ -69,10 +73,6 @@
-
-=======
->>>>>>> master
 Ville Bergholm, Josh Izaac, Angus Lowe, Johannes Jakob Meyer, Oluwatobi Ogunbayo, Maria Schuld
```
Not sure what this is doing on master

co9olguy commented on doc/introduction/templates.rst:
```diff
@@ -228,6 +228,7 @@ The following initialization functions are available:
 .. autosummary::
     :nosignatures:
 
+<<<<<<< HEAD
```
more merge issues here

mariaschuld replied:
Very sorry, removed and made sure that there are no merge-like things in the entire path...

co9olguy commented on tests/test_init.py:
```diff
@@ -71,47 +71,45 @@
-       ]
-ALL_SHP = [[intf, intf, wrs, wrs, wrs, intf, intf, wrs, wrs, wrs, wrs],
-           [(n_if,), (n_if,), (n_wires,)]]
+SHAPE = [(repeat, n_rots), (repeat, n_wires, 3), intf_block, intf_block, wire_block, wire_block,
```
The names are better, but I still find myself struggling to parse easily what these elements refer to.
Suggestion: Can we recast it so the shapes are right next to the corresponding inits? (essentially a hand-written `zip`)
```
INIT_AND_SHAPES = [[(qml.init.random_layers_normal, rnd_rpt_nrml), (repeat, n_rots)], 
                   [(qml.init.strong_ent_layers_normal, rpt_nrml), (repeat, n_wires, 3)],
                   ...
                  ]
```
From that, we could easily make separate `INIT` and `SHAPE` objects programmatically using list comprehensions

mariaschuld replied:
I will address this in the PR #419, since I changed the signature slightly there. Want to avoid merge conflicts.

mariaschuld commented on tests/test_templates.py:
```diff
@@ -83,13 +86,15 @@
-               (AngleEmbedding, [[1., 2.]])
-              ]
+               (AngleEmbedding, [[1., 2.]]),
+               ]
```
Yes! Johannes PR changed this.

mariaschuld commented on tests/test_templates.py:
```diff
@@ -18,13 +18,14 @@
 
-* When adding a new interface, try to import it and extend the fixture ``interfaces``.
+* When adding a new interface, try to import it and extend the fixture ``interfaces``. Add interface to
+  TestGradientIntegration tests.
```
I clarified this statement.

mariaschuld commented:
@josh146 , any changes requested from your side?

mariaschuld commented:
@co9olguy I addressed all your requests.

josh146 commented on doc/introduction/templates.rst:
```diff
@@ -152,8 +228,24 @@ randomly initialized arrays for the trainable parameters.
+    ~pennylane.init.cvqnn_layers_phi_a_uniform
+    ~pennylane.init.cvqnn_layers_phi_a_normal
+    ~pennylane.init.cvqnn_layers_kappa_uniform
+    ~pennylane.init.cvqnn_layers_kappa_normal
```
Instead of listing all these hear, can we just link to the `qml.init` api page, where they are all listed as well? Avoids duplication and saves with maintenance.

mariaschuld replied:
I propose we do this in a separate docs PR that adapts the docs to the templates redesign.

josh146 commented on doc/introduction/templates.rst:
```diff
@@ -164,7 +256,12 @@ randomly initialized arrays for the trainable parameters.
+    ~pennylane.init.interferometer_phi_uniform
+    ~pennylane.init.interferometer_phi_normal
+    ~pennylane.init.interferometer_varphi_uniform
+    ~pennylane.init.interferometer_varphi_normal
```
Ditto here.

co9olguy commented:
> @co9olguy I addressed all your requests.

Yep already approved from my side
